src/bpf/bpfctl.c

```diff
@@ -0,0 +1,215 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <getopt.h>
+
+#include "build.h"
+#include "format-table.h"
+#include "format-util.h"
+#include "main-func.h"
+#include "parse-argument.h"
+#include "pretty-print.h"
+#include "varlink.h"
+#include "verbs.h"
+
+static JsonFormatFlags arg_json_format_flags = JSON_FORMAT_OFF;
+
+typedef struct BpfProgData BpfProgData;
+typedef struct BpfMapData BpfMapData;
+
+struct BpfProgData {
+        char *name;
+        uint64_t type;
+        char *memlock;
+        char *tag;
+        char *map_names;
+        uint64_t run_time_ns;
+        uint64_t run_cnt;
+};
+
+struct BpfMapData {
+        char *name;
+        uint64_t type;
+        char *memlock;
+};
+
+static void bpf_prog_data_done(BpfProgData *p) {
+        assert(p);
+
+        p->name = mfree(p->name);
+        p->memlock = mfree(p->memlock);
+        p->tag = mfree(p->tag);
+        p->map_names = mfree(p->map_names);
+}
+
+static void bpf_map_data_done(BpfMapData *p) {
+        assert(p);
+
+        p->name = mfree(p->name);
+        p->memlock = mfree(p->memlock);
+}
+
+static int query_bpf_progs_and_maps(void) {
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        _cleanup_(varlink_unrefp) Varlink *vl = NULL;
+        _cleanup_(json_variant_unrefp) JsonVariant *o = NULL;
+        JsonVariant *i = NULL, *m = NULL, *v = NULL;
+        int r;
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
+        if (r < 0)
+                return log_debug_errno(r, "Unable to connect to /run/systemd/bpf/io.systemd.Bpf: %m");
```

Comment:
In applications we don't need to do debug logging so you can use `log_error_errno()` here instead.